api/common/sharing/utils.go

```diff
@@ -12,44 +12,236 @@ import (
+		}
+		return util.HandleInstanceSharingError(util.ErrNoResultsForReferenceSelector, "")
+	}
+	if results.Len() > 1 && len(parameters) == 0 {
+		return util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	}
+	if results.Len() > 1 {
+		// there is more than one shared instance that meets your criteria
+		return util.HandleInstanceSharingError(util.ErrMultipleReferenceSelectorResults, "")
+	}
+	return nil
+}
+
+func getSharedInstancesByTenant(ctx context.Context, repository storage.Repository, tenantIdentifier, tenantID string) (types.ObjectList, error) {
+	sharedInstances, err := getSharedInstances(ctx, repository, tenantIdentifier, tenantID)
+	if err != nil {
+		return nil, util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	return sharedInstances, nil
+}
+
+func validateParameters(parameters map[string]gjson.Result) error {
+	// when provisioning reference by instance id, we only allow one parameter:
+	_, byID := parameters[instance_sharing.ReferencedInstanceIDKey]
+	if byID && len(parameters) > 1 {
+		return util.HandleInstanceSharingError(util.ErrInvalidReferenceSelectors, "")
+	}
+	return nil
+}
+
+func getSharedInstances(ctx context.Context, repository storage.Repository, tenantIdentifier string, tenantID string) (types.ObjectList, error) {
+	referencePlan, _ := types.PlanFromContext(ctx)
+	queryParams := map[string]interface{}{
+		"tenant_identifier": tenantIdentifier,
+		"tenant_id":         tenantID,
+		"offering_id":       referencePlan.ServiceOfferingID,
 	}
 
-	return referencedInstanceID.String(), nil
+	objectList, err := repository.QueryForList(ctx, types.ServiceInstanceType, storage.QueryForSharedInstances, queryParams)
+	return objectList, err
+}
+
+func filterInstancesBySelectors(ctx context.Context, repository storage.Repository, instances types.ObjectList, parameters map[string]gjson.Result, smaap bool) (types.ServiceInstances, error) {
+	var err error
+	var filteredInstances types.ServiceInstances
+	for i := 0; i < instances.Len(); i++ {
+		if filteredInstances.Len() > 1 {
+			break
+		}
+		// selectors: true or false -> the entire map should be true in order to pass
+		selectors := make(map[string]bool)
+		selectorVal, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+
+		// "*" for any shared instance of the tenant, if parameter not passed, set true for any shared instance
+		if !exists || selectorVal.String() == "*" {
+			selectors["id"] = true
+		}
+
+		instance := instances.ItemAt(i).(*types.ServiceInstance)
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceInstanceNameSelector]
+		if exists {
+			selectors["name"] = instance.Name == selectorVal.String()
+		}
+
+		selectorVal, exists = parameters[instance_sharing.ReferencePlanNameSelector]
+		if exists {
+			selectors["service_plan_id"], err = foundSharedInstanceWithPlan(ctx, repository, instance, selectorVal.String(), smaap)
+			if err != nil {
```

Comment:
you can do it with one query and not query per instance
first bring all plans of shared instances and just look in the result

Reply:
in that case, we better use a different named query with inner join on service plans

that brings us back to the question of single named query or many...

Reply:
It brings us back to the question of a single `named_query` or many.
I think when we have a `plan_name_selector` we should run another `named_query` with inner join on service plans.
Instead of comparing two lists.

What do you think?

Reply:
no, just bring corresponding plans after you brought the shared instances
and look in there instead of query per instance
if you have 5 instances and user asked by plan you'll end with 5 queries

Reply:
done, the query moved to the start of the selector function.
so if you pass the plan selector, we get the plan with its `offeringID`, if found a plan we compare it.

single request instead of multiple for each instance.